geode-core/src/main/java/org/apache/geode/internal/net/filewatch/FileWatchingX509ExtendedKeyManager.java

```diff
@@ -0,0 +1,137 @@
+
+  private static final Logger logger = LogService.getLogger();
+  private static final ConcurrentHashMap<Path, FileWatchingX509ExtendedKeyManager> instances =
+      new ConcurrentHashMap<>();
+
+  private final AtomicReference<X509ExtendedKeyManager> keyManager = new AtomicReference<>();
+  private final Path keyStorePath;
+  private final ThrowingSupplier<KeyManager[]> keyManagerSupplier;
+
+  @VisibleForTesting
+  FileWatchingX509ExtendedKeyManager(Path keystorePath,
+      ThrowingSupplier<KeyManager[]> keyManagerSupplier,
+      ExecutorService executor) {
+    this.keyStorePath = keystorePath;
+    this.keyManagerSupplier = keyManagerSupplier;
+
+    loadKeyManager();
+
+    executor.submit(new FileWatcher(this.keyStorePath, this::loadKeyManager));
+  }
+
+  /**
+   * Returns a {@link FileWatchingX509ExtendedKeyManager} for the given path. A new instance will be
+   * created only if one does not already exist for that path.
+   *
+   * @param path The path to the key store file
+   * @param supplier A supplier which returns an {@link X509ExtendedKeyManager}
+   */
+  public static FileWatchingX509ExtendedKeyManager forPath(Path path,
+      ThrowingSupplier<KeyManager[]> supplier) {
+    return instances.computeIfAbsent(path,
+        (Path p) -> new FileWatchingX509ExtendedKeyManager(p, supplier, newSingleThreadExecutor()));
+  }
+
+  @Override
+  public String chooseClientAlias(String[] strings, Principal[] principals, Socket socket) {
+    return keyManager.get().chooseClientAlias(strings, principals, socket);
+  }
+
+  @Override
+  public String chooseEngineClientAlias(String[] strings, Principal[] principals,
+      SSLEngine sslEngine) {
+    return keyManager.get().chooseEngineClientAlias(strings, principals, sslEngine);
+  }
+
+  @Override
+  public String chooseEngineServerAlias(String s, Principal[] principals, SSLEngine sslEngine) {
+    return keyManager.get().chooseEngineServerAlias(s, principals, sslEngine);
+  }
+
+  @Override
+  public String chooseServerAlias(String s, Principal[] principals, Socket socket) {
+    return keyManager.get().chooseServerAlias(s, principals, socket);
+  }
+
+  @Override
+  public X509Certificate[] getCertificateChain(String s) {
+    return keyManager.get().getCertificateChain(s);
+  }
+
+  @Override
+  public String[] getClientAliases(String s, Principal[] principals) {
+    return keyManager.get().getClientAliases(s, principals);
+  }
+
+  @Override
+  public PrivateKey getPrivateKey(String s) {
+    return keyManager.get().getPrivateKey(s);
+  }
+
+  @Override
+  public String[] getServerAliases(String s, Principal[] principals) {
+    return keyManager.get().getServerAliases(s, principals);
+  }
+
+  private void loadKeyManager() {
+    KeyManager[] keyManagers;
+    try {
+      keyManagers = keyManagerSupplier.get();
+    } catch (Throwable t) {
+      throw new RuntimeException("Unable to load KeyManager", t);
+    }
```

Comment:
Usually we try to avoid catching VirtualMachineErrors and its subclasses. For example if OutOfMemoryError is thrown and caught here, there's actually no guarantee it will be thrown again/elsewhere even if the JVM is OOM.

It's usually more appropriate to use InternalGemFireException or GemFireSecurityException (both are RuntimeExceptions):
```
throw new InternalGemFireException("Unable to load KeyManager", t);
```

Reply:
Thanks! I will change this.